internal/provider/providers/vultr/provider.go

```diff
@@ -0,0 +1,195 @@
+package vultr
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/netip"
+	"net/url"
+
+	"github.com/qdm12/ddns-updater/internal/models"
+	"github.com/qdm12/ddns-updater/internal/provider/constants"
+	"github.com/qdm12/ddns-updater/internal/provider/errors"
+	"github.com/qdm12/ddns-updater/internal/provider/headers"
+	"github.com/qdm12/ddns-updater/internal/provider/utils"
+	"github.com/qdm12/ddns-updater/pkg/publicip/ipversion"
+)
+
+type Provider struct {
+	domain     string
+	owner      string
+	ipVersion  ipversion.IPVersion
+	ipv6Suffix netip.Prefix
+	token      string
+	ttl        uint32
+}
+
+func New(data json.RawMessage, domain, owner string,
+	ipVersion ipversion.IPVersion, ipv6Suffix netip.Prefix) (
+	provider *Provider, err error) {
+
+	var providerSpecificSettings struct {
+		Token string `json:"token"`
+		TTL   uint32 `json:"ttl"`
+	}
+	err = json.Unmarshal(data, &providerSpecificSettings)
+	if err != nil {
+		return nil, fmt.Errorf("json decoding provider specific settings: %w", err)
+	}
+
+	err = validateSettings(domain, owner, providerSpecificSettings.Token, providerSpecificSettings.TTL)
+	if err != nil {
+		return nil, fmt.Errorf("validating provider specific settings: %w", err)
+	}
+
+	return &Provider{
+		domain:     domain,
+		owner:      owner,
+		ipVersion:  ipVersion,
+		ipv6Suffix: ipv6Suffix,
+		token:      providerSpecificSettings.Token,
+		ttl: 		providerSpecificSettings.TTL,	
+	}, nil
+}
+
+func validateSettings(domain, owner, token string, ttl uint32) (err error) {
+	err = utils.CheckDomain(domain)
+	if err != nil {
+		return fmt.Errorf("%w: %w", errors.ErrDomainNotValid, err)
+	}
+
+	switch {
+	case owner == "":
+		return fmt.Errorf("%w", errors.ErrOwnerNotSet)
+	case token == "":
+		return fmt.Errorf("%w", errors.ErrAPIKeyNotSet)
```

Comment:
Isn't there an `ErrTokenNotSet`? If not, please create one 🙏 Otherwise it's rather confusing for the user, since it's a `token` and the error talks about an api key.

Reply:
This is a bit confusing, the token is actually a bearer token, but from their settings page they're asking for an API key, I think APIKeyNotSet makes more sense as that's what the user is asked to send.

Reply:
I've updated to be token not set, let me know what you prefer though 🙏🏽

Reply:
<img width="536" alt="image" src="https://github.com/user-attachments/assets/97838f39-4e58-4bfc-8baf-9a3fcd6bd364">

Reply:
Oh ok, let's match what vultr has, so instead rename that token field to `apikey` and use that previous apikey not set error. The user shouldn't (and probably doesn't) care about the implementation details such as how it's sent to the API (bearer token).

Reply:
Fixed, take a look pls 🙏🏽

Reply:
Thanks! I also love how they say "Personal Access Token" and also "API key"... I guess it doesn't matter too much 😄